features/26-mfa/api.md

```diff
@@ -0,0 +1,436 @@
+
+interface AuthenticateExtraOptions {
+  requestBearerToken?: boolean;
+}
+
+interface AuthenticateResult {
+  bearerToken?: string;
+}
+
+interface TOTPAuthenticateOptions {
+  authenticatorID?: string;
+  otp: string;
+}
+
+async function authenticateWithTOTP(options: TOTPAuthenticateOptions, extraOptions?: AuthenticateExtraOptions): Promise<AuthenticateResult>;
+
+interface OOBAuthenticateOptions {
+  authenticatorID?: string;
+  code: string;
+}
+
+async function authenticateWithOOB(options: OOBAuthenticateOptions, extraOptions?: AuthenticateExtraOptions): Promise<AuthenticateResult>;
+
+async function authenticateWithRecoveryCode(code: string, extraOptions?: AuthenticateExtraOptions): Promise<AuthenticateResult>;
+
+async function authenticateWithBearerToken(token: string): Promise<void>;
+
+// Authenticator management
+
+type Authenticator = TOTPAuthenticator | OOBAuthenticator | RecoveryCodeAuthenticator | BearerTokenAuthenticator;
+
+interface TOTPAuthenticator {
+  id: string;
+  type: "totp";
+  activatedAt: Date;
+  displayName: string;
+}
+
+interface OOBAuthenticator {
+  id: string;
+  type: "oob";
+  activatedAt: Date;
+  channel: "sms";
+  phone: string;
+}
+
+interface RecoveryCodeAuthenticator {
+  id: string;
+  type: "recovery_code";
+  createdAt: Date;
+}
+
+interface BearerTokenAuthenticator {
+  id: string;
+  type: "bearer_token";
+  createdAt: Date;
+  expireAt: Date;
+}
+
+interface RegenerateRecoveryCodeResult {
+  recoveryCodes: string[];
+}
+
+async function listAuthenticators(): Promise<Authenticator[]>;
+async function deleteAuthenticator(authenticatorID: string): Promise<void>;
+async function regenerateRecoveryCode(): Promise<RegenerateRecoveryCodeResult>;
+```
+
+# Use Cases
+
+## Registration
+
+```typescript
+try {
+  await skygear.auth.login("user@example.com", "password");
+} catch (e) {
+  if (skygear.isMFARequiredError(e)) {
+    navigateToRegisterMFAScreen();
+    return;
+  }
+  // Handle any other error.
+}
+```
+
+```typescript
+// Present an UI to let the user to choose authenticator.
+// Suppose the user chose TOTP.
+
+// Present AN UI to let the user to optionally name the TOTP authenticator.
+
+const displayName = textInput.value;
+const { authenticatorID, secret } = await skygear.auth.mfa.createNewTOTP(displayName);
+
+// Present to the secret to the user and let the user
+// to add the secret to their TOTP application, such as Authy and Google Authenticator.
+
+// Present an UI to instruct the user to input the OTP.
+const otp = textInput.value;
+const { recoveryCodes } = await skygear.auth.mfa.activateTOTP(authenticatorID, otp);
+```
+
+```typescript
+// Present an UI to let the user to choose authenticator.
+// Suppose the user chose OOB.
+
+// Present an UI to instruct the user to input a phone number.
+const e164number = textInput.value;
+const { authenticatorID } = await skygear.auth.mfa.createNewOOB({ phone: e164number });
+
+// Present an UI to instruct the user to check SMS.
```

Comment:
I assume `createNewOOB` will automatically do `triggerOOB`?

Reply:
Yes. You're right.